File: androidApp/src/main/java/com/oddinstitute/crossplatformsvgparser/android/DrawView.kt
package com.oddinstitute.crossplatformsvgparser.android

import android.content.Context
import android.graphics.Canvas
import android.graphics.DashPathEffect
import android.graphics.Paint
import android.view.View
import com.oddinstitute.crossplatformsvgparser.to_refactor.Artwork
import com.oddinstitute.crossplatformsvgparser.operators.toFloatArray


class DrawView(context: Context) : View(context)
{
    var artwork: Artwork? = null
    var paint: Paint = Paint()

    fun redraw(art: Artwork)
    {
        this.artwork = art
        this.artwork?.let {
            for (obj in it.objects)
                obj.makePath()
        } // FIXME -> Is this needed. Maybe it causes a double rendering
        this.invalidate()
    }


    override fun onDraw(canvas: Canvas)
    {
        artwork?.let {
            drawArtwork(canvas, it)
        }
    }
}

fun DrawView.drawArtwork(canvas: Canvas, artwork: Artwork)
{ // drawing all paths
    for (obj in artwork.objects)
    { // fill


        if (obj.shapeAttr.fillColor != null ) {

            styleFillPaint(obj)

            obj.path?.let {
                canvas.drawPath(it.makeCgPath(), paint)
            }
        }

        if (obj.shapeAttr.strokeWidth > 0f)
        {
            styleStrokePaint(obj)
            obj.path?.let {
                canvas.drawPath(it.makeCgPath(), paint)
            }
//            canvas.drawPath(obj.myPath.path, paint)
        }
    }

}

fun DrawView.styleFillPaint(obj: com.oddinstitute.crossplatformsvgparser.objects.Object)
{
    paint.style = Paint.Style.FILL



    obj.shapeAttr.filColorApplied?.let {
        paint.color = Utils.myColorToArgb(it)
    }

    // paint.color = Color.RED
//    paint.strokeCap = Utils.svgLineCapToType(obj.shapeAttr.strokeLineCap)
//    paint.strokeJoin = Utils.svgLineJoinToType(obj.shapeAttr.strokeLineJoin)

    paint.pathEffect = null
    paint.clearShadowLayer()
}


// todo important
fun DrawView.styleStrokePaint(obj: com.oddinstitute.crossplatformsvgparser.objects.Object)
{
    obj.shapeAttr.strokeWidth.let { width ->
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = width


        obj.shapeAttr.strokeColorApplied?.let {
            paint.color = Utils.myColorToArgb(it)
        }

        paint.strokeCap = Utils.svgLineCapToType(obj.shapeAttr.strokeLineCap)
        paint.pathEffect = null
        paint.clearShadowLayer()


        // FIXME -> We don't know how this float array works
        // it might have many entires, we are only currnently readn the first
        obj.shapeAttr.dashArray?.let {

            val f = it.toFloatArray()
            if (it.isNotEmpty()) paint.pathEffect = DashPathEffect(floatArrayOf(f[0], f[0]), 0f)

        }

        paint.strokeJoin = Utils.svgLineJoinToType(obj.shapeAttr.strokeLineJoin)

    }
}

File: iosApp/iosApp/DrawView.swift
import SwiftUI
import shared


//struct MyView: View {
//    var body: some View {
//
//        Text("Hello, world!")
//            .padding()
//    }
//}

struct DrawView: View {
    
    var artwork: Artwork
    
    func getStyle () -> StrokeStyle
    {
        var style = StrokeStyle()
        style.lineCap = .butt
        style.lineWidth = 20
        
        return style
    }
    
    
//    func styleFillPaint(obj: Object) -> FillStyle
//    {
//        var style = FillStyle ()
//
//        if let it = obj.shapeAttr.fillColor {
//
//        }
//
//        // style.isEOFilled = obj.shapeAttr.
//
//        obj.shapeAttr.filColorApplied?.let {
//            paint.color = Utils.myColorToArgb(it)
//        }
//    }
//
    
    
    var body: some View {
        
        
        

        for obj in artwork.objects {
            (obj as! Object).makePath()
            
            
            if ((obj as! Object).shapeAttr.fillColor != nil)
            {
                // there is fill color
                
                // here set the style
                // styleFillPaint(obj)
                // and draw the path
                // it.makeCgPath()
            }
            
            if ((obj as! Object).shapeAttr.strokeWidth > 0)
            {
                // here set the stroke style
               //  styleStrokePaint(obj)
                
                // and draw stroke path
                // it.makeCgPath()
            }
            
            
        }
        
        
        
        return ObjectShape(artwork: artwork)
            .stroke(Color.orange, style: getStyle())
    }
}


//right now, we are mainly in the Draw View



struct ObjectShape: Shape
{
    var artwork: Artwork?
    
    func path(in rect: CGRect) -> Path {
        
 
        
        
        
        var path = Path()
        path.move(to: CGPoint(x: rect.size.width, y: 0))
        path.addLine(to: CGPoint(x: rect.size.width, y: rect.size.width))
        path.addLine(to: CGPoint(x: 0, y: rect.size.width))
        path.addLine(to: CGPoint(x: 0, y: 0))
        path.closeSubpath()
        
        
        
        return path
    }
}

